### PR TITLE
LinkedList

### PR DESCRIPTION
**Pull request by AndrewLinkii**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,48 +3,170 @@
...
+
+    private void addToEmpty(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        headNode = newNode;
+        tailNode = newNode;
+        size++;
+    }
+
+    private void addFirst(T value) {
+        if (size == 0) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: do we need this `if` here:
if (size == 0) { 
addToEmpty(value);
        }

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,48 +3,170 @@
...
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("wrong index"
+                    + "\n" + "Mast be from 0 to "
+                    + (size() - 1));
+        }
+        if (size() == 0) {
+            addToEmpty(value);
+        } else if (index == 0) {
+            addFirst(value);
+        } else if (index == size()) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
        } else if (index == size() || size() == 0) {
```
and then remove 
if (size() == 0) {
            addToEmpty(value);
        }

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,48 +3,170 @@
...
+            add(value);
+        } else {
+            Node<T> prevNode = getNodeFromIndex(index);
+            Node<T> nextNode = getNodeFromIndex(index);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: what is the difference between line 91 and 92?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,48 +3,170 @@
...
     @Override
     public T get(int index) {
-        return null;
+        return getNodeFromIndex(index).value;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: do we need this checking here?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,48 +3,170 @@
...
+        size--;
+        return element;
+
     }
 
     @Override
     public T remove(T t) {
+        for (int index = 0; index < size(); index++) {
+            if (getNodeFromIndex(index).value.equals(t)) {
+                T searchedItem = getNodeFromIndex(index).value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]: T searchedItem = getNodeFromIndex(index).value;
                remove(index);
                return searchedItem;
-->
return  remove(index);

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 1 above)
Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: No, but it like insurance

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 2 above)
Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: total agree

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 4 above)
Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: agree

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 1 above)
Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: Without checking i get NullPointerException

Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: Teste failed

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 3 above)
Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: A serious mistake, it’s strange that the test passed

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 5 above)
Comment (AndrewLinkii) [reply src/main/java/core/basesyntax/MyLinkedList.java]: I'm sorry, i do not understand
Index it's the local variable

Comment (Sofasmile) [reply src/main/java/core/basesyntax/MyLinkedList.java]: ми можемо замінити цих три рядки на один return remove(index);

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,48 +3,162 @@
...
+                    + "\n" + "Mast be from 0 to "
+                    + (size() - 1));
+        }
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size() || size() == 0) {
+            add(value);
+        } else {
+            Node<T> prevNode = getNodeFromIndex(index - 1);
+            Node<T> nextNode = getNodeFromIndex(index + 1);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/MyLinkedList.java]:
```suggestion
            Node<T> nextNode = prevNode.next;
```

**Pull request closed without merging**